keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -372,6 +430,32 @@ Based on the data collected from user clusters, vast majority (> 99%) of the k8s
 
 The current Endpoints API only includes a boolean state (Ready vs. NotReady) on individual endpoint. However, according to pod life cycle, there are more states (e.g. Graceful Termination, ContainerReary). In order to represent additional states other than Ready/NotReady,  a status structure is included for each endpoint. More condition types can be added in the future without compatibility disruptions. As more conditions are added, different consumer (e.g. different kube-proxy implementations) will have the option to evaluate the additional conditions. 
 
+- #### Why not use a CRD?
+
+**1. Protobuf is more efficient** Currently CRDs don't support protobuf. In our
+testing, a protobuf watch is approximately 5x faster than a JSON watch. We used
+pprof to profile 2 versions of kube-proxy using EndpointSlices and running on 2
+different nodes in a 150 node cluster as it scaled up to 15k endpoints. Over the
+15 minute window, kube-proxy with JSON used 17% more CPU time, with the
+difference in `StreamWatcher.receive` accounting for all of that. With protobuf
+enabled, that function took 1/5th the time of the JSON implementation.
+
+**2. Validation is too complex** Validation of addresses relies on addressType,
+something that would be difficult, maybe impossible, to recreate with OpenAPI
+validations. Additionally, there are a number of validations currently in use
+that are able to reuse the same validations used elsewhere for Services or
+Endpoints such as `IsDNS1123Label` and `IsValidIP`. Although these could be
+recreated with OpenAPI validations, the error messages would not be as helpful
+and we would lose the consistency in messaging from the related resources.
+
+**3. EndpointSlices are required for the API Server to be accessible** In an
```

Comment:
Yeah, ok.  I'd still like to break this but maybe not yet.

Reply:
Using CRD sort of required a certain bootstrap sequence:
1. CRD ensured
2. controller startup. 

For CRD consumers, there is no good way to wait for a CRD to showed up. Most implementations just die and wait for the restart to try again.

Reply:
we have to address this in general, anyway